File: q26-50/q36.py
# 2020.08.08
# Problem Statement:
# https://leetcode.com/problems/valid-sudoku/

class Solution:
    # help function
    # return if a list has duplication (invalid)
    def isDuplicate(input_list):
        for element in input_list:
            if element != "." and input_list.count(element) > 1:
                return True
        return False
    
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        
        # check rows
        for i in range(0, 9):
            if Solution.isDuplicate(board[i]):
                return False
            
        # check columns
        # create a new board to store columns in original board as rows
        temp_board = []
        temp_list = []
        for i in range(0, 9):
            for j in range(0, 9):
                temp_list.append(board[j][i])            
            temp_board.append(temp_list)
            temp_list = []
        # check valid
        for i in range(0, 9):
            if Solution.isDuplicate(temp_board[i]):
                return False   
            
        # check 3*3 boxes
        # create a new board to store 3*3 boxes in original board as rows
        temp_board = []
        for i in range(0, 3):
            for j in range(0, 3):
                for x in range(3*i, 3*i+3):
                    for y in range(3*j, 3*j+3):
                        temp_list.append(board[x][y])
                temp_board.append(temp_list)
                temp_list = []
        # check valid        
        for i in range(0, 9):
            if Solution.isDuplicate(temp_board[i]):
                return False 
                
        return True

File: Amazon/q36.java
// 2022.05.28
// Problem Statement:
// https://leetcode.com/problems/valid-sudoku/

// idea: use stack to check duplication in rows/cols/3*3 cells
class Solution {
    public boolean isValidSudoku(char[][] board) {
        // check row, col and 3*3 cells
        // row
        for (int i=0; i<9; i++) {
            HashSet<Character> row_set = new HashSet<> ();
            for (int j=0; j<9; j++) {
                if (board[i][j]!='.') {
                    if (row_set.contains(board[i][j])) return false;
                    row_set.add(board[i][j]);
                }
            }
        }
        // col
        for (int j=0; j<9; j++) {
            HashSet<Character> col_set = new HashSet<> ();
            for (int i=0; i<9; i++) {
                if (board[i][j]!='.') {
                    if (col_set.contains(board[i][j])) return false;
                    col_set.add(board[i][j]);
                }
            }
        }
        // 3*3 cells
        for (int cnt=0; cnt<9; cnt++) {
            HashSet<Character> cell_set = new HashSet<> ();
            for (int i=3*(cnt/3); i<3*(cnt/3)+3; i++) {
                for (int j=3*(cnt%3); j<3*(cnt%3)+3; j++) {
                    if (board[i][j]!='.') {
                        if (cell_set.contains(board[i][j])) return false;
                        cell_set.add(board[i][j]);
                    }
                }
            }
        }
        return true;
    }
}